File: 141. Linked List Cycle/Unordered_set/LinkedListCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == nullptr){
            return false;
        }


        unordered_set<ListNode*> visited;
        ListNode* temp = head;


        while (temp != nullptr){
            if(visited.find(temp) != visited.end()){
                return true;
            }


            visited.insert(temp);
            temp = temp->next;
        }
        return false;
    }
};

File: 141. Linked List Cycle/Unordered_set/LinkedListCycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if head is None:
            return False

        visited = set()
        temp = head

        while temp != None:
            if temp in visited:
                return True
            
            visited.add(temp)
            temp = temp.next

        return False
        